### ISSUE TITLE
[Design] Allow connector participating in query planning

### ISSUE THREAD

**Issue by hellium01** [prestodb/presto#12368]
The detail design is shared here: https://docs.google.com/document/d/1GJatB0MQ9OcUS0m0K5O9J3AGRa5yIO0bKLghDxXdOho/edit?usp=sharing
### **Introduction**
Presto provides flexible connector interface to allow reading data from many different data source, which makes it a perfect solution to consolidate data into a big data lake. However, today the freedom we give connector is very limited and connector can only act as a data source. In many cases, the upstream system we connect to has some level of compute capacity and being able to utilize those capacity and do calculation close to data storage can help us improve the efficiency of many different connectors.  Many connector such as MySQL/Cassandra/… support a subset of SQL dialect and we can execute aggregation even join in those data sources with presto receiving only partial results. 

Assuming we have a connector that can read data from a fast source that supports partial computes and a slow source that only can serves as data storage. We want to rewrite the plan into a union over aggregation result of slow source and tableScan that returns intermediate results from fast data source.

### Problem of current architecture

Currently, in presto, we convert AST into Plan tree after analyze step. We don’t separate LogicalPlan and ExecutionPlan so PlanNode today contains a lot of optional information we filled in for execution. As the plan converge close to physical plan, more and more information is added into PlanNode. This causes our plan to be very complex and have a lot of dependency, which contradict our principle of connector: hiding as much as implementation details of engine from connector as possible.

### Proposed change to allow connector participate in planning
We propose to add a new connector based optimizer which happens at a fixed point of time of planning (before AddExchange which marks a starting point of transition from logical planning to physical planning). 

In SPI, we will ask connector to provide  ConnectorRuleFactory which returns a list of rules the connector supports. The connector provided optimization rule has following interface:
```
Interface Rule 
{
     boolean match(IR subtree);
     Optional<IR> apply(IR subtree);
}
```
In the optimizer, we collect Traits such as ConnectorId for each subtree root and record it in memo. We then iterate the plan tree from bottom up, and attempt to:
Convert subtree to a simplified intermediate representation (IR) we can send to connector
Check if the subtree belong to a certain connector
Match the subtree generated IR with connector provided rules
Matched rule will attempt to rewrite the subtree and return a rewritten subtree to engine
Once a rewritten IR is received, we convert that back to planNode and replace the subtree.

### Proposed IR
We can represent the relation with two type of operations:

- Operation on column sets (Table).
- Operation on columns.

Any relation such as project/filter/tableScan/aggregate can be represented as tableExpression. Leaf table expression can be tableScan or values … etc. Branch nodes can be considered as function over table(column sets). 
Each table expression, project/filter action can be considered as operations over column (channel). Such operations over column can be represented by ColumnExpression, which only have following node types:

- FunctionCall
- Lambda 
- Input
- Constant
- InputReference (either from symbol, channel, or column handle)

###  Push down common operations in connectors
Under above framework, connector can easily rewrite the subtree to pushdown the oeprations into table scan:

**Predicate push down:**
We will try to attempt to rewrite subtree of  the column expression that represents the predicate. Whatever we can rewrite into connector operation will be added into tableLayout, and the leftover part will be returned as a filter node.

**Project push down:**

For each assignment, we attempt to rewrite the subtree of column expression. Those can be understand by connector will be replaced with a columnhandle that represents the read operation. Those left over operation tree will remain in project node. 

**Aggregation push down:**
Aggregation push down is somewhat similar to project pushdown. Except, we rewrite the aggregation node to accept intermediate type instead of raw data.

**Comment by hellium01** [comment]
cc @highker , @jessesleeping , @wenleix , @arhimondr

**Comment by highker** [comment]
Other PRs/issues that push down specific optimizations through spi: #11454, #2508, #11326, #5547, #10064, #5547, #10064.

**Comment by highker** [comment]
solved by #12920
